nncf/common/graph/graph.py

```diff
@@ -148,17 +148,23 @@ def __init__(
         self.parallel_input_port_ids = parallel_input_port_ids
 
     def __str__(self):
-        return str(self.from_node) + " -> " + str(self.tensor_shape) + " -> " + str(self.to_node)
+        return f"{self.from_node}:{self.output_port_id} -> {self.tensor_shape} -> {self.to_node}:{self.input_port_id}"
 
     def __hash__(self):
-        return hash(str(self))
+        return hash(
+            (
+                self.from_node,
+                self.to_node,
+                self.input_port_id,
+                self.output_port_id,
+                tuple(self.tensor_shape),
+                self.dtype,
+                tuple(self.parallel_input_port_ids),
+            )
+        )
 
     def __eq__(self, other):
-        return (
-            self.from_node == other.from_node
-            and self.to_node == other.to_node
-            and self.tensor_shape == other.tensor_shape
-        )
+        return self.__dict__ == other.__dict__
```

Comment:
Same concern as for hash

Reply:
Why? Iterating over a dict and comparing is O(N) where N is the number of fields

Reply:
My concern is that before there were 3 comparisons, but now there are more -- since `self.__dict__` has more values.

Reply:
https://github.com/openvinotoolkit/nncf/pull/2107/files/d950c520a6cdeefaccfb56cad4ffbd1c5ceff03a#r1324503677